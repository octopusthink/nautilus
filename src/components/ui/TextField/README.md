A TextField is an editable input that enables users to enter and interact with free-form text data.

```jsx
<TextField label="Feed me sentences!" />

<TextField defaultValue="Lorem ipsum..." label="I have a default value." />
```

## Usage

### Use a `TextField` for...

Use a `TextField` when you want users to be able to enter or edit free-form text.

### Don't use a `TextField` for...

- Selecting from a predefined list of options. Use a `DropDownSelect` instead.
- Choosing a date. Use a `DatePicker` instead.
- Selecting a colour. Use a `ColorPicker` instead.
- Any text the user isn't able to edit themselves. Use a `Paragraph` instead.

## Appearance

### Labelling

Every `TextField` has a visible **label**, aligned directly above the input they describe. Labels are short, succinct, and clearly communicate what text the user should enter.

**Hint text** is always visible and appears beneath the label. Use a hint when you want to provide more context or additional guidance. Hints can be more verbose than a label and should provide helpful tips or instructions that will be relevant for the majority of users, like how their information will be used, or where to find a reference number.

```jsx
<TextField label="Fluffy things" hint="List the fluffiest things you can think of, ideally in alphabetical order." placeholder="e.g.: Kittens, puppies, and ponies." />
```

A **placeholder** can be used to provide examples of what sort of information to enter. Placeholder text disappears once the user starts typing and should never be used for critical information. Use placeholder text sparingly. Placeholder text should be used to illustrate examples rather than provide instructions.

```jsx
<TextField label="Fluffy things" placeholder="e.g.: Kittens, puppies, and ponies." />
```

### Size

By default, a `TextField` spans the full width of its parent element in a single line. This size can be adjusted to provide a visual cue as to how much data the input expects.

The size of a `TextField` should give users a visual cue as to how much text is expected. If you want the user to enter two characters, restrict the size to two characters. If you'd like to encourage users to enter a longer piece of text, use a `multiline`.

```jsx
<TextField label="Postcode" size={7} />

<TextField label="I am a multiline input, useful for entering a lot of tex." multiline />
```

### Optional fields

A `TextField` can be marked as optional if it's data that some users need to complete, but others wouldn't. If a `TextField` is marked as optional because it's not strictly required information, consider removing the input altogether.

The text string to mark an optional field is customisable in your theme. **Do not** rely on symbols (asterisks, stars, exclamation marks, etc) or colour alone to communicate the difference between required and non-required fields, since not all users will immediately understand what you're trying to say.

```jsx
<TextField label="I am optional" optional />
```

## Interaction

### Focus

When an input is focussed, it gains additional prominence.

### Disabled

Use a disabled input to indicate that an input isn't ready yet, or relies on another input elsewhere in the screen.

```jsx
<TextField label="Disabled input" size={100} disabled defaultValue="You don't need to worry about me." />
```

### Error handling

Error validation should happen inline in order to give users immediate feedback. Don't start validation until the user has stopped typing inside the input, to avoid frustration.

```jsx
<TextField label="Trouble in paradise" size={14} error="Please fix me!" />
```

### Autocomplete

The [autocomplete attribute](https://cloudfour.com/thinks/autofill-what-web-devs-should-know-but-dont/) helps users complete forms more quickly by allowing them to use browser tools to autofill frequently-entered information. Users complete forms up to 30% faster when autocomplete is implemented well.

Use [standard values](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#autofill) where appropriate to match the `TextField`'s intent. If a field should not be autocompleted, set the value to `no`.

### Default values

Where possible, programmatically determine default values to `TextFields` to make it easier for users to complete a form. When providing defaults, make sure those defaults are user-editable in case of an error or false positive.

## Voice & Tone

**Label text** should be short and snappy. Aim for one to three words in total. Don't use conversational labels or any punctuation at the end of the label. ("Name" vs "Tell me your name.")

**Hint text** should use full sentences and can be more verbose. ("Enter your full postal address, including your postal code.")

**Placeholder text** should only be used for examples; not for instructions or extra context. Include “e.g.” before examples and write in sentence case.

## Accessibility

### Use visible labels

A text input should always have a visible label. Screen reader users rely on these to know what to input.

Labels must be visible when an input gains focus, and must be announced to screen readers on focus.

**Never** use a placeholder as a label for an input. These disappear once the user starts typing text and so should only be used for examples to underline a user's understanding of what they need to enter. Use a label to label the input, and use help text to provide supplementary information.

### Only ask questions you really need to ask

Limit the number of inputs (of any type) on a single page. When possible, improve user focus by only asking one question per page and splitting longer forms across multiple pages.

Only ask users for information you really require. There's a mental cost to every additional input you ask a user to fill in ([https://en.wikipedia.org/wiki/Decision_fatigue](https://en.wikipedia.org/wiki/Decision_fatigue)). If the majority of your inputs are optional, you should consider removing those you don't need.

Be very careful asking for sensitive data (titles, gender, ethnicity, age, etc). Collecting this data requires careful privacy and data retention considerations and may make users uncomfortable. It can also be used to inadvertently reinforce stereotypes. Ask yourself why you are collecting data, and, if you can, avoid it entirely.

### Use the correct type

Mobile devices and some assistive technologies adapt the software keyboard to the current task. This helps users with mobility, vision, and cognitive issues to enter information more easily.

If the field is for an email address, for instance, use `type="email"`; this will often enable a software keyboard with an `@` symbol more easily accessible.

### Order labels and help text correctly

Ensure the helper text that appears under an input is read when an assistive technology user stops at an input using ARIA.

### Each input should have a unique ID

Good news: Nautilus does this for you already! If you _do_ supply a custom `id`, `TextField` will detect this and assign things like the `<label>`'s `htmlFor` the correct value.

### Don't be prescriptive about names

Don't separate names into different inputs (e.g. "first name", "last name", etc.). [Not all cultures have the same sorts of naming conventions](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/), and imposing your culture's naming scheme onto others can make it harder for users to enter their preferred name. Unless absolutely necessary, use a single "name" field.
