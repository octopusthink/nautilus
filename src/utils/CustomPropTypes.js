import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
