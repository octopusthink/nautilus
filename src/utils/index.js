export { default as CustomPropTypes } from './CustomPropTypes';
