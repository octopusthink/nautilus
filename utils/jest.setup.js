// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
