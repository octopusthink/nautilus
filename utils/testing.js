// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
