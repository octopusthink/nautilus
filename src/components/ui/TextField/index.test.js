import React, { createRef } from 'react';

import { axe, fireEvent, muteConsole, render } from 'utils/testing';

import TextField from '.';

describe('TextField', () => {
  it('should render a <label> and <div><input></div> by default', () => {
    const { container } = render(<TextField label="Hello" />);

    expect(container.children[0].tagName).toEqual('LABEL');
    expect(container.children[1].tagName).toEqual('DIV');
    expect(container.children[1].children[0].tagName).toEqual('INPUT');
  });

  it('should not output an empty <label> if one is not supplied', () => {
    muteConsole({ times: 1, type: 'error' });
    const { container } = render(<TextField />);

    expect(container.children[0].tagName).toEqual('DIV');
    expect(container.children[0].children[0].tagName).toEqual('INPUT');
  });

  it('will render `children` after the <label> and <div><input></div> elements', () => {
    const { container } = render(
      <TextField label="Hello">
        <span>Why is this even here? I dunno</span>
      </TextField>,
    );

    expect(container.children[0].tagName).toEqual('LABEL');
    expect(container.children[1].tagName).toEqual('DIV');
    expect(container.children[1].children[0].tagName).toEqual('INPUT');
    expect(container.children[2].tagName).toEqual('SPAN');
  });

  it('should render a <label> and <div><textarea></div> when `multiline` is `true`', () => {
    const { container } = render(<TextField label="Hello" multiline />);

    expect(container.children[0].tagName).toEqual('LABEL');
    expect(container.children[1].tagName).toEqual('DIV');
    expect(container.children[1].children[0].tagName).toEqual('TEXTAREA');
  });

  it('should set rows to 4 by default', () => {
    const { getByTestId } = render(<TextField data-testid="input" label="Hello" multiline />);

    expect(getByTestId('input').getAttribute('rows')).toEqual('4');
  });

  it('should not set a `max-width` on the input element without a `size` prop', () => {
    const { container } = render(<TextField id="testInput" label="Hello" />);

    expect(container).toMatchSnapshot();
  });

  it('should set a `max-width` on the input element when `size` is set', () => {
    const { container } = render(<TextField id="testInput" label="Hello" size={10} />);

    expect(container).toMatchSnapshot();
  });

  it('should set a `max-width` on the textarea element when `size` is set', () => {
    const { container } = render(<TextField id="testInput" label="Hello" multiline size={10} />);

    expect(container).toMatchSnapshot();
  });

  it('should not change a generated ID on subsequent renders', () => {
    const { container, rerender } = render(<TextField label="Hello" />);

    const inputId = container.querySelector('input').getAttribute('id');
    expect(inputId).toBeDefined();

    rerender(<TextField size={4} />);
    const reRenderedInputId = container.querySelector('input').getAttribute('id');

    expect(reRenderedInputId).toEqual(inputId);
  });

  it('should not change a generated error message ID on subsequent renders', () => {
    const error = <div data-testid="errorMessage">Something went wrong!</div>;
    const { getByTestId, rerender } = render(<TextField error={error} label="Hello" />);

    const errorId = getByTestId('errorMessage').getAttribute('id');

    rerender(<TextField error={error} size={7} />);
    const reRenderedErrorId = getByTestId('errorMessage').getAttribute('id');

    expect(reRenderedErrorId).toEqual(errorId);
  });

  it('should use the existing `id` prop on an error component', () => {
    const error = (
      <div data-testid="errorMessage" id="cool-error-one" label="Hello">
        Something went wrong!
      </div>
    );
    const { getByTestId } = render(<TextField data-testid="input" error={error} label="Hello" />);

    const ariaLabel = getByTestId('input').getAttribute('aria-errormessage');
    const errorId = getByTestId('errorMessage').getAttribute('id');
    expect(ariaLabel).toEqual('cool-error-one');
    expect(errorId).toEqual('cool-error-one');
  });

  it('should call `onBlur` function prop', () => {
    const onBlur = jest.fn();
    const { getByTestId } = render(
      <TextField data-testid="blurInput" label="Hello" onBlur={onBlur} />,
    );

    fireEvent.blur(getByTestId('blurInput'));

    expect(onBlur).toHaveBeenCalled();
  });

  it('should call `onFocus` function prop', () => {
    const onFocus = jest.fn();
    const { getByTestId } = render(
      <TextField data-testid="focusInput" label="Hello" onFocus={onFocus} />,
    );

    fireEvent.focus(getByTestId('focusInput'));

    expect(onFocus).toHaveBeenCalled();
  });

  it('should match styles', () => {
    const { container } = render(<TextField label="Hello" id="myTextField" />);

    expect(container).toMatchSnapshot();
  });

  it('should hide styles when unstyled prop is set', () => {
    const { container } = render(<TextField label="Hello" id="myTextField" unstyled />);

    expect(container).toMatchSnapshot();
  });

  it("shouldn't have a margin when `noMargin` is true", () => {
    const { container } = render(<TextField label="Hello" noMargin id="myTextField" />);

    expect(container).toMatchSnapshot();
  });

  it("shouldn't have a margin when `noMargin` is true and error exists", () => {
    const error = <div id="errorMessage">Something went wrong!</div>;
    const { container } = render(
      <TextField label="Hello" error={error} noMargin id="myTextField" />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should output an <Icon> when the signifierIcon prop is used', () => {
    const { container, getByTestId } = render(
      <TextField
        __signifierIconId="signifierIcon"
        signifierIcon="phone"
        data-testid="focusInput"
        label="Hello"
        id="myTestId"
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(getByTestId('signifierIcon').tagName).toEqual('SPAN');
    expect(getByTestId('signifierIcon').firstChild.tagName).toEqual('svg');
  });

  it('should focus the text field when the signifier icon is clicked', () => {
    const ref = createRef();

    const { getByTestId } = render(
      <TextField
        __signifierIconId="signifierIcon"
        signifierIcon="phone"
        data-testid="focusInput"
        label="Hello"
        id="myTestId"
        ref={ref}
      />,
    );

    const focusSpy = jest.spyOn(ref.current, 'focus');

    // The onClick event gets attached to the SVG child element, so make sure
    // that's what we click on.
    fireEvent.click(getByTestId('signifierIcon').firstChild);

    expect(focusSpy).toHaveBeenCalled();
  });

  it('should output an <Icon> when the actionIcon prop is used', () => {
    const { container, getByTestId } = render(
      <TextField
        __actionIconId="arnoldIcon"
        actionIcon="phone"
        data-testid="focusInput"
        label="Hello"
        id="myTestId"
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(getByTestId('arnoldIcon').tagName).toEqual('SPAN');
    expect(getByTestId('arnoldIcon').firstChild.tagName).toEqual('svg');
  });

  it('should allow the user to click on the actionIcon', () => {
    const onClick = jest.fn();

    const { getByTestId } = render(
      <TextField
        __actionIconId="arnoldIcon"
        actionIcon="phone"
        actionIconOnClick={onClick}
        data-testid="focusInput"
        label="Hello"
        id="myTestId"
      />,
    );

    // The onClick event gets attached to the SVG child element, so make sure
    // that's what we click on.
    fireEvent.click(getByTestId('arnoldIcon').firstChild);

    expect(onClick).toHaveBeenCalled();
  });

  it('should forward refs', () => {
    const ref = createRef();

    render(<TextField label="Hello" ref={ref} />);

    expect(ref.current).not.toBeNull();
    expect(ref.current.tagName).toEqual('INPUT');
  });

  describe('accessibility', () => {
    it('should set `aria-errormessage` prop on input component when `error` prop is set', () => {
      const { getByTestId } = render(
        <TextField data-testid="input" error="Bad input!" label="Hello" />,
      );

      const ariaLabel = getByTestId('input').getAttribute('aria-errormessage');
      expect(ariaLabel).toBeDefined();
    });

    it("should use the error message's ID for `aria-errormessage`", () => {
      const error = <div data-testid="errorMessage">Something went wrong!</div>;
      const { getByTestId } = render(<TextField data-testid="input" error={error} label="Hello" />);

      const ariaLabel = getByTestId('input').getAttribute('aria-errormessage');
      const errorId = getByTestId('errorMessage').getAttribute('id');
      expect(ariaLabel).toEqual(errorId);
    });

    it('should require a label prop', () => {
      muteConsole({ times: 1, type: 'error' });
      render(<TextField label={null} />);

      expect(global.console.error.mock.calls[0][0]).toMatch(
        'Warning: Failed prop type: The prop `label` is marked as required in `TextField`, but its value is `null`.',
      );
    });

    it('should pass aXe tests', async () => {
      const { container } = render(<TextField label="Hello" />);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });
  });
});
