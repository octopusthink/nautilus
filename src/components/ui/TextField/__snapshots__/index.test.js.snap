// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should hide styles when unstyled prop is set 1`] = `
.emotion-2 {
  position: relative;
}

<div>
  <label
    class="emotion-0"
    for="myTextField"
  >
    Hello
  </label>
  <div
    class="emotion-2"
  >
    <input
      class="emotion-0"
      id="myTextField"
      required=""
      type="text"
    />
  </div>
</div>
`;

exports[`TextField should match styles 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  color: #3b3f45;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 50%;
  -ms-flex: 1 0 50%;
  flex: 1 0 50%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 0.4rem;
}

.emotion-0:active {
  color: #cd2f83;
}

.emotion-0:focus {
  color: #cd2f83;
}

.emotion-2 {
  position: relative;
}

.emotion-1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  background: #ffffff;
  border-radius: 0;
  border: 2px solid #3b3f45;
  box-sizing: border-box;
  color: #3b3f45;
  display: block;
  margin: 0;
  outline: none;
  padding: 1.2rem;
  position: relative;
  -webkit-transition: box-shadow 200ms;
  transition: box-shadow 200ms;
  width: 100%;
  margin: 0 0 3.2rem;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1:focus {
  color: #181b1c;
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #666c76;
}

.emotion-1::-moz-placeholder {
  color: #666c76;
}

.emotion-1:-ms-input-placeholder {
  color: #666c76;
}

.emotion-1::placeholder {
  color: #666c76;
}

<div>
  <label
    class="emotion-0"
    for="myTextField"
  >
    Hello
  </label>
  <div
    class="emotion-2"
  >
    <input
      class="emotion-1"
      id="myTextField"
      required=""
      type="text"
    />
  </div>
</div>
`;

exports[`TextField should not set a \`max-width\` on the input element without a \`size\` prop 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  color: #3b3f45;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 50%;
  -ms-flex: 1 0 50%;
  flex: 1 0 50%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 0.4rem;
}

.emotion-0:active {
  color: #cd2f83;
}

.emotion-0:focus {
  color: #cd2f83;
}

.emotion-2 {
  position: relative;
}

.emotion-1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  background: #ffffff;
  border-radius: 0;
  border: 2px solid #3b3f45;
  box-sizing: border-box;
  color: #3b3f45;
  display: block;
  margin: 0;
  outline: none;
  padding: 1.2rem;
  position: relative;
  -webkit-transition: box-shadow 200ms;
  transition: box-shadow 200ms;
  width: 100%;
  margin: 0 0 3.2rem;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1:focus {
  color: #181b1c;
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #666c76;
}

.emotion-1::-moz-placeholder {
  color: #666c76;
}

.emotion-1:-ms-input-placeholder {
  color: #666c76;
}

.emotion-1::placeholder {
  color: #666c76;
}

<div>
  <label
    class="emotion-0"
    for="testInput"
  >
    Hello
  </label>
  <div
    class="emotion-2"
  >
    <input
      class="emotion-1"
      id="testInput"
      required=""
      type="text"
    />
  </div>
</div>
`;

exports[`TextField should set a \`max-width\` on the input element when \`size\` is set 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  color: #3b3f45;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 50%;
  -ms-flex: 1 0 50%;
  flex: 1 0 50%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 0.4rem;
}

.emotion-0:active {
  color: #cd2f83;
}

.emotion-0:focus {
  color: #cd2f83;
}

.emotion-2 {
  position: relative;
}

.emotion-1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  background: #ffffff;
  border-radius: 0;
  border: 2px solid #3b3f45;
  box-sizing: border-box;
  color: #3b3f45;
  display: block;
  margin: 0;
  outline: none;
  padding: 1.2rem;
  position: relative;
  -webkit-transition: box-shadow 200ms;
  transition: box-shadow 200ms;
  width: 100%;
  margin: 0 0 3.2rem;
  max-width: 9em;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1:focus {
  color: #181b1c;
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #666c76;
}

.emotion-1::-moz-placeholder {
  color: #666c76;
}

.emotion-1:-ms-input-placeholder {
  color: #666c76;
}

.emotion-1::placeholder {
  color: #666c76;
}

<div>
  <label
    class="emotion-0"
    for="testInput"
  >
    Hello
  </label>
  <div
    class="emotion-2"
  >
    <input
      class="emotion-1"
      id="testInput"
      maxlength="10"
      required=""
      type="text"
    />
  </div>
</div>
`;

exports[`TextField should set a \`max-width\` on the textarea element when \`size\` is set 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  color: #3b3f45;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 50%;
  -ms-flex: 1 0 50%;
  flex: 1 0 50%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 0.4rem;
}

.emotion-0:active {
  color: #cd2f83;
}

.emotion-0:focus {
  color: #cd2f83;
}

.emotion-2 {
  position: relative;
}

.emotion-1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  background: #ffffff;
  border-radius: 0;
  border: 2px solid #3b3f45;
  box-sizing: border-box;
  color: #3b3f45;
  display: block;
  margin: 0;
  outline: none;
  padding: 1.2rem;
  position: relative;
  -webkit-transition: box-shadow 200ms;
  transition: box-shadow 200ms;
  width: 100%;
  margin: 0 0 3.2rem;
  max-width: 9em;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1:focus {
  color: #181b1c;
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #666c76;
}

.emotion-1::-moz-placeholder {
  color: #666c76;
}

.emotion-1:-ms-input-placeholder {
  color: #666c76;
}

.emotion-1::placeholder {
  color: #666c76;
}

<div>
  <label
    class="emotion-0"
    for="testInput"
  >
    Hello
  </label>
  <div
    class="emotion-2"
  >
    <textarea
      class="emotion-1"
      id="testInput"
      maxlength="10"
      required=""
      rows="4"
    />
  </div>
</div>
`;

exports[`TextField shouldn't have a margin when \`noMargin\` is true 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  color: #3b3f45;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 50%;
  -ms-flex: 1 0 50%;
  flex: 1 0 50%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 0.4rem;
}

.emotion-0:active {
  color: #cd2f83;
}

.emotion-0:focus {
  color: #cd2f83;
}

.emotion-2 {
  position: relative;
}

.emotion-1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  background: #ffffff;
  border-radius: 0;
  border: 2px solid #3b3f45;
  box-sizing: border-box;
  color: #3b3f45;
  display: block;
  margin: 0;
  outline: none;
  padding: 1.2rem;
  position: relative;
  -webkit-transition: box-shadow 200ms;
  transition: box-shadow 200ms;
  width: 100%;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1:focus {
  color: #181b1c;
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #666c76;
}

.emotion-1::-moz-placeholder {
  color: #666c76;
}

.emotion-1:-ms-input-placeholder {
  color: #666c76;
}

.emotion-1::placeholder {
  color: #666c76;
}

<div>
  <label
    class="emotion-0"
    for="myTextField"
  >
    Hello
  </label>
  <div
    class="emotion-2"
  >
    <input
      class="emotion-1"
      id="myTextField"
      required=""
      type="text"
    />
  </div>
</div>
`;

exports[`TextField shouldn't have a margin when \`noMargin\` is true and error exists 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  color: #3b3f45;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 50%;
  -ms-flex: 1 0 50%;
  flex: 1 0 50%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 0 0.4rem;
}

.emotion-0:active {
  color: #cd2f83;
}

.emotion-0:focus {
  color: #cd2f83;
}

.emotion-2 {
  position: relative;
}

.emotion-1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  background: #ffffff;
  border-radius: 0;
  border: 2px solid #3b3f45;
  box-sizing: border-box;
  color: #3b3f45;
  display: block;
  margin: 0;
  outline: none;
  padding: 1.2rem;
  position: relative;
  -webkit-transition: box-shadow 200ms;
  transition: box-shadow 200ms;
  width: 100%;
  border-color: #c26767;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1:focus {
  color: #181b1c;
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #666c76;
}

.emotion-1::-moz-placeholder {
  color: #666c76;
}

.emotion-1:-ms-input-placeholder {
  color: #666c76;
}

.emotion-1::placeholder {
  color: #666c76;
}

<div>
  <label
    class="emotion-0"
    for="myTextField"
  >
    Hello
  </label>
  <div
    class="emotion-2"
  >
    <input
      aria-errormessage="errorMessage"
      class="emotion-1"
      id="myTextField"
      required=""
      type="text"
    />
    <div
      id="errorMessage"
    >
      Something went wrong!
    </div>
  </div>
</div>
`;
